internal/k8s/reconciler/deployer.go

```diff
@@ -0,0 +1,121 @@
+	return &GatewayDeployer{
+		client:   config.Client,
+		consulCA: config.ConsulCA,
+		sdsHost:  config.SDSHost,
+		sdsPort:  config.SDSPort,
+		logger:   config.Logger,
+	}
+}
+
+func (d *GatewayDeployer) Deploy(ctx context.Context, namespace string, config apigwv1alpha1.GatewayClassConfig, gateway *gw.Gateway) error {
+	if err := d.ensureServiceAccount(ctx, config, gateway); err != nil {
+		return err
+	}
+
+	if err := d.ensureDeployment(ctx, namespace, config, gateway); err != nil {
+		return err
+	}
+
+	return d.ensureService(ctx, config, gateway)
+}
+
+func (d *GatewayDeployer) ensureServiceAccount(ctx context.Context, config apigwv1alpha1.GatewayClassConfig, gateway *gw.Gateway) error {
+	// Create service account for the gateway
+	if serviceAccount := config.ServiceAccountFor(gateway); serviceAccount != nil {
+		if err := d.client.EnsureServiceAccount(ctx, gateway, serviceAccount); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (d *GatewayDeployer) ensureDeployment(ctx context.Context, namespace string, config apigwv1alpha1.GatewayClassConfig, gateway *gw.Gateway) error {
+	deployment := d.Deployment(namespace, config, gateway)
+	mutated := deployment.DeepCopy()
+	if updated, err := d.client.CreateOrUpdateDeployment(ctx, mutated, func() error {
+		mutated = apigwv1alpha1.MergeDeployment(deployment, mutated)
+		return d.client.SetControllerOwnership(gateway, mutated)
+	}); err != nil {
+		return fmt.Errorf("failed to create or update gateway deployment: %w", err)
+	} else if updated {
+		if d.logger.IsTrace() {
+			data, err := json.MarshalIndent(mutated, "", "  ")
+			if err == nil {
+				d.logger.Trace("created or updated gateway deployment", "deployment", string(data))
+			}
+		}
+	}
+
+	return nil
+}
+
+func (d *GatewayDeployer) ensureService(ctx context.Context, config apigwv1alpha1.GatewayClassConfig, gateway *gw.Gateway) error {
+	if service := d.Service(config, gateway); service != nil {
+		mutated := service.DeepCopy()
+		if updated, err := d.client.CreateOrUpdateService(ctx, mutated, func() error {
+			mutated = apigwv1alpha1.MergeService(service, mutated)
+			return d.client.SetControllerOwnership(gateway, mutated)
+		}); err != nil {
+			return fmt.Errorf("failed to create or update gateway service: %w", err)
+		} else if updated {
+			if d.logger.IsTrace() {
+				data, err := json.MarshalIndent(mutated, "", "  ")
+				if err == nil {
+					d.logger.Trace("created or updated gateway service", "service", string(data))
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func (d *GatewayDeployer) Deployment(namespace string, config apigwv1alpha1.GatewayClassConfig, gateway *gw.Gateway) *apps.Deployment {
+	deploymentBuilder := builder.NewGatewayDeployment(gateway)
+	deploymentBuilder.WithSDS(d.sdsHost, d.sdsPort)
+	deploymentBuilder.WithClassConfig(config)
+	deploymentBuilder.WithConsulCA(d.consulCA)
+	deploymentBuilder.WithConsulGatewayNamespace(namespace)
+	return deploymentBuilder.Build()
```

Comment:
#nit This feels odd for consuming a builder pattern. I'm used to seeing

```suggestion
	return builder.NewGatewayDeployment(gateway).
	    WithSDS(d.sdsHost, d.sdsPort).
	    WithClassConfig(config).
	    WithConsulCA(d.consulCA).
            WithConsulGatewayNamespace(namespace).
            Build()
```

Reply:
Agree, and I'm fine with changing a bit of the usage of the builder pattern down the line, but don't quite consider it in scope right now for this pass.